Support Channel sending data which does not implement Copy

StevenJiang1110 opened the pull request:
The implementation is trivial since the low-level ring buffer supports pushing and poping non-Copy data types. The push and pop method reuse much code from current read/write methods.

tatetian commented:
Could you explain the use case for passing non-Copy objects through this channel?

StevenJiang1110 commented:
> Could you explain the use case for passing non-Copy objects through this channel?

I want to use `Channel` to support multicast in netlink. A message will be sent to multiple receivers through multiple Channels. A message can be viewed as `Arc<[u8]>`. If the channel can support sending types which does not implement `Copy`, we can simple clone `Arc<[u8]>` instead of really copy bytes. 

```rust
let message = copy_message_from_user();

let (producer1, consumer1) = Channel::new().split();
let (producer2, cousumer2) = Channel::new().split();
let (producer3, consumer3) = Channel::new().split();

producer1.push(message.clone());
producer2.push(message.clone());
producer3.push(message);
```

tatetian reviewed (approved):
LGTM. Thanks for the contribution!
